### PR TITLE
[Chore] Bump katana version to 4.0.0-beta

### PR DESCRIPTION
**Pull request by albertogiunta**

### CODE REVIEW

File: Podfile [hunk 1]
```diff
@@ -6,12 +6,13 @@ use_frameworks!
 target 'Tempura' do
   platform :ios, '9.0'
   podspec
+  
+  pod 'Katana', :git => 'https://github.com/BendingSpoons/katana-swift.git', :branch => 'feature/documentation-and-tests'
```
Comment (cipolleschi) [inline Podfile]: remember to remove this!

File: Tempura/Core/ViewController.swift [hunk 2]
```diff
@@ -226,15 +226,25 @@ open class ViewController<V: ViewControllerModellableView & UIView>: UIViewContr
...
-    self.store.dispatch(action)
+  /// Shortcut to the non-generic dispatch function.
+  open func dispatch(_ dispatchable: Dispatchable) {
+    self.store.dispatch(dispatchable)
   }
   
   /// Shortcut to the dispatch function.
+  open func dispatch<T: Dispatchable>(_ dispatchable: T) {
+    _ = self.store.dispatch(dispatchable)
+  }
```
Comment (smaramba) [inline Tempura/Core/ViewController.swift]: This is not needed as, given there are no return values, it's not possible to invoke this method if `dispatch(_ dispatchable: Dispatchable)` is already defined.

This is a small playground to prove it.
```swift
protocol State {
  init()
}


class Promise<Value> {
  
}

public protocol Dispatchable {}

protocol AnySideEffect: Dispatchable {}

protocol SideEffect: AnySideEffect {
  associatedtype ReturnValue
}

protocol AnyStateUpdater: Dispatchable {}

protocol StateUpdater: AnyStateUpdater {
  associatedtype StateType: State
}


struct SE1: SideEffect {
  typealias ReturnValue = Int
}

struct SU1: StateUpdater {
  typealias StateType = AppState
}

protocol AnyStore: class {
  var anyState: State { get }
  
  @discardableResult
  func dispatch<T: Dispatchable>(_ dispatchable: T) -> Promise<Void>
  
  @discardableResult
  func dispatch<T: SideEffect>(_ dispatchable: T) -> Promise<T.ReturnValue>
  
  @discardableResult
  func dispatch(_ dispatchable: Dispatchable) -> Promise<Any>
}


class PartialStore<S: State>: AnyStore {
  var state: S
  
  var anyState: State {
    return self.state
  }
  
  init(state: S) {
    self.state = state
  }
  
  @discardableResult
  func dispatch<T: Dispatchable>(_ dispatchable: T) -> Promise<Void> {
    print("#1")
    fatalError()
  }
  
  @discardableResult
  func dispatch<T: SideEffect>(_ dispatchable: T) -> Promise<T.ReturnValue> {
    print("#2")
    fatalError()
  }
  
  @discardableResult
  func dispatch(_ dispatchable: Dispatchable) -> Promise<Any> {
    print("#3")
    fatalError()
  }
}

struct AppState: State {}

let store = PartialStore<AppState>(state: AppState())

//store.dispatch(SE1()) //#3
//store.dispatch(SU1()) //#3

//let p = store.dispatch(SU1()) //#3

//let p: Promise<Void> = store.dispatch(SU1()) //#1

//store.dispatch(SE1()) //#3

//let p: Promise<Int> = store.dispatch(SE1()) //#2

//let p: Promise<Any> = store.dispatch(SE1()) //#3

//let c  = store.dispatch(SU1()) //#3

class VC<S: State> {
  var store: PartialStore<S>
  
  init(store: PartialStore<S>) {
    self.store = store
  }
  
  func dispatch(_ dispatchable: Dispatchable) {
    self.store.dispatch(dispatchable)
  }
  
  func dispatch<T: Dispatchable>(_ dispatchable: T) {
    self.store.dispatch(dispatchable)
  }
}

//let c: Promise<Void> = store.dispatch(SU1()) //#1

let vc = VC<AppState>(store: store)
//vc.dispatch(SU1()) // #3
//vc.dispatch(SE1()) // #3
// here there is no possibility to invoke #1 or #2
```

File: Tempura/Navigation/NavigationActions.swift [hunk 3]
```diff
@@ -28,36 +28,10 @@ public struct Navigate: AnySideEffect {
...
-    self.animated = animated
-    self.context = context
-  }
-  
-  /// The side effect of the action, look into [Katana](https://github.com/BendingSpoons/katana-swift)
-  /// to know what a `SideEffect` is.
-  public func sideEffect(currentState: State, previousState: State, dispatch: @escaping StoreDispatch, dependencies: SideEffectDependencyContainer) {
-    guard let dependencies = dependencies as? NavigationProvider else { fatalError("DependenciesContainer must conform to `NavigationProvider`") }
-    let _ = dependencies.navigator.changeRoute(newRoute: self.route, animated: self.animated, context: self.context).void
+    return Void()
```
Comment (smaramba) [inline Tempura/Navigation/NavigationActions.swift]: I'd probably prefer `return ()`.

File: Tempura/Navigation/NavigationActions.swift [hunk 4]
```diff
@@ -100,54 +74,10 @@ public struct Show: AnySideEffect {
...
-    where K: RawRepresentable, K.RawValue == RouteElementIdentifier {
-      self.init(identifierToShow.rawValue, animated: animated, context: context)
-  }
-  
-  /// The side effect of the action, look into [Katana](https://github.com/BendingSpoons/katana-swift)
-  /// to know what a `SideEffect` is.
-  public func sideEffect(currentState: State, previousState: State, dispatch: @escaping StoreDispatch, dependencies: SideEffectDependencyContainer) {
-    guard let dependencies = dependencies as? NavigationProvider else { fatalError("DependenciesContainer must conform to `NavigationProvider`") }
-    let _ = dependencies.navigator.show(self.identifiersToShow, animated: self.animated, context: self.context).void
+    return Void()
```
Comment (smaramba) [inline Tempura/Navigation/NavigationActions.swift]: I'd probably prefer `return ()`.

File: Tempura/Navigation/NavigationActions.swift [hunk 5]
```diff
@@ -191,52 +121,9 @@ public struct Hide: AnySideEffect {
   
   /// The side effect of the action, look into [Katana](https://github.com/BendingSpoons/katana-swift)
   /// to know what a `SideEffect` is.
-  public func sideEffect(_ context: AnySideEffectContext) throws {
+  public func anySideEffect(_ context: AnySideEffectContext) throws -> Any {
     guard let dependencies = context.anyDependencies as? NavigationProvider else { fatalError("DependenciesContainer must conform to `NavigationProvider`") }
     try await(dependencies.navigator.hide(self.identifierToHide, animated: self.animated, context: self.context, atomic: self.atomic))
+    return Void()
```
Comment (smaramba) [inline Tempura/Navigation/NavigationActions.swift]: I'd probably prefer `return ()`.

File: .gitignore [hunk 6]
```diff
@@ -20,3 +20,7 @@ Carthage/
 /build
 /Demo/UITests/
 TempuraHelpers.xcscheme
+/Tempura.xcodeproj/xcshareddata/xcschemes/TempuraTesting.xcscheme
+/Tempura.xcodeproj/xcshareddata/xcschemes/Tempura.xcscheme
+/Tempura.xcodeproj/xcshareddata/xcschemes/Demo.xcscheme
+/Tempura.xcodeproj/project.pbxproj
```
[the remark this answers is not in the document]
Comment (albertogiunta) [reply .gitignore]: ah ecco alright, i was wondering why it was tracked in the first place, thanks 🙏

[the remark this answers is not in the document]
Comment (smaramba) [reply .gitignore]: 🙌

File: Tempura/Navigation/NavigationActions.swift [hunk 7]
```diff
@@ -191,52 +123,41 @@ public struct Hide: AnySideEffect {
...
-  /// Initializes and return a Hide action.
-  public init<K>(_ identifierToHide: K, animated: Bool = false, context: Any? = nil)
-    where K: RawRepresentable, K.RawValue == RouteElementIdentifier {
-      self.init(identifierToHide.rawValue, animated: animated, context: context)
+}
+
+extension SideEffectContext {
+  public func dispatch<RSE: TempuraSideEffect>(_ dispatchable: RSE) -> Promise<Void> {
+    return self.anyDispatch(dispatchable).void
   }
```
Comment (smaramba) [inline Tempura/Navigation/NavigationActions.swift]: Are any of these helpers used in this `NavigationActions.swift`? If not, I think we should move these in a separate file under `Utilities` as a file named `SideEffectContext+Utils.swift`.

Comment (smaramba) [reply Tempura/Navigation/NavigationActions.swift]: Do we really need these helpers? I feel like their place should be Katana, the only reason they are here I suppose is because of the `TempuraSideEffect` thing, my question is, do we need this `TempuraSideEffect` protocol?

File: Tempura/Navigation/NavigationActions.swift [hunk 8]
```diff
@@ -191,52 +123,41 @@ public struct Hide: AnySideEffect {
...
-    let _ = dependencies.navigator.hide(self.identifierToHide, animated: self.animated, context: self.context, atomic: self.atomic).void
+}
+
+extension ViewController {
+  public func __unsafeDispatch<RSE: TempuraSideEffect>(_ dispatchable: RSE) -> Promise<Void> {
+    return self.store.dispatch(dispatchable)
+  }
+
+  public func __unsafeAwaitDispatch<RSE: TempuraSideEffect>(_ dispatchable: RSE) throws {
+    return try await(self.store.dispatch(dispatchable))
```
Comment (smaramba) [inline Tempura/Navigation/NavigationActions.swift]: Same as before. Do we need it?

File: Tempura/Navigation/NavigationActions.swift (hunk 7 above)
Comment (FEC-bendingspoons) [reply Tempura/Navigation/NavigationActions.swift]: We can move them, I initially put them here because they are semantically related only to things inside this file

Comment (FEC-bendingspoons) [reply Tempura/Navigation/NavigationActions.swift]: Also yes, we need them to have a clean katana dispatch exposed to the app. Since tempura dispatchable inherit directly from `AnySideEffect` (and not `SideEffect`) the only dispatch that you can use to with them is `anyDispatch(_:) -> Promise<Any>` which is `poco elegante`.

File: Tempura/Navigation/NavigationActions.swift (hunk 8 above)
Comment (FEC-bendingspoons) [reply Tempura/Navigation/NavigationActions.swift]: same as above

File: Tempura/Navigation/NavigationActions.swift (hunk 7 above)
Comment (smaramba) [reply Tempura/Navigation/NavigationActions.swift]: I got your point now! Then this is the right place for the extensions, but we need to rename `TempuraSideEffect` to `NavigationSideEffect` to express the fact that we expect a TempuraSideEffect to be used only in the context of navigation actions.

File: Tempura/UITests/UITests.swift [hunk 9]
```diff
@@ -31,14 +31,14 @@ public enum UITests {
    */
   
   
-  public struct VCScreenSnapshot<VC: AnyViewController> {
```
Comment (smaramba) [inline Tempura/UITests/UITests.swift:34]: Why do we want to change the typing of UITests in this PR?
The new typing doesn't contain the link between the ViewController and the View, I could create a VCScreenSnapshot with a ViewController and a UIView that are not related in any way. We used the `AnyViewController` to express the fact that the VC and the View need to be linked.
